### ISSUE TITLE
Environment variables to identify Lambda runtime environment

### ISSUE THREAD

**Issue by kbalk** [spulec/moto#4638]
This is a followup question to the PR ["Feature: Custom cloudformation resources"](https://github.com/spulec/moto/pull/4512) and perhaps related to the issue ["switch server mode to proxies"](https://github.com/spulec/moto/issues/1108).

The lambda submitted as a test case for that PR issued API calls to the EC2 service.  Part of the fix required the lambda to specify the endpoint_url to any boto3 services it invoked, i.e.:

`ec2 = boto3.client("ec2", endpoint_url="http://host.docker.internal:5000")`

That lambda (or the test case) can be invoked in different environmentss:

- against AWS itself (no endpoint_url value needed),
- when LocalStack and moto are started up from the command line (the endpoint_url of "http://host.docker.internal:5000" works great here),
- when LocalStack and moto are started within a docker container and then start up the lambda within that container.

I have not been able to figure out the appropriate endpoint_url for the last case.  I also don't have a way to determine which of the three situations I'm in.  If I relied on LocalStack, I could use the environment variable `LOCALSTACK_HOSTNAME` to identify when I'm running against AWS versus localstack.  However, if I only use moto endpoints, I don't have an equivalent environment variable.

Since my test case uses Terraform, I've tried passing variables declared in the Terraform modules, but I still can't distinguish between the three scenarios and I'm trying to avoid hardcoding the moto port.

Do you have any suggestions as to how I might solve this problem?  I could live with hardcoding the moto port if I knew the appropriate endpoint for the third case above and if I had a way of knowing which endpoint I should be using.

**Comment by bblommers** [comment]
Half of the time involved in creating the custom resources feature was figuring out that exact same problem!

Have a look at the settings-file - the `moto_server_host()` should give you the right value for the hostname.

https://github.com/spulec/moto/blob/master/moto/settings.py#L53:L78

Note that the port is set by the `moto_server`-script, so you shouldn't have to set this env manually.

**Comment by kbalk** [comment]
Your suggestions look promising ... I'll give them a try this weekend.

**Comment by kbalk** [comment]
When I created this issue, I was thinking of logic that I could add inside the lambda that could detect the environment.  Although `moto/moto/settings.py` has the functions I could use for determining the port and server, I have two problems:

1.  I still need an external environment variable injected into lambda to determine if I'm running in a mock environment.  I've used LOCALSTACK_HOSTNAME before for that purpose.  Moto isn't injecting an environment variable that I can use.  I could pass a Terraform variable to the Terraform module that then sets environment variables before invoking the Lambda, but I was hoping to avoid that.
2. Within the Lambda, I can't import the `moto/moto/settings.py` functions.  The following lines in the lambda 

```
if os.environ.get("MOCK_AWS"):
    from moto.settings import moto_server_port, moto_server_host
```
gives me a stack trace:

```
Error on request:
Traceback (most recent call last):
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/werkzeug/serving.py", line 323, in run_wsgi
    execute(self.server.app)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/werkzeug/serving.py", line 312, in execute
    application_iter = app(environ, start_response)
  File "/home/kbalk/plus3it/moto/moto/server.py", line 239, in __call__
    return backend_app(environ, start_response)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/kbalk/plus3it/moto/moto/core/utils.py", line 147, in __call__
    result = self.callback(request, request.url, {})
  File "/home/kbalk/plus3it/moto/moto/core/responses.py", line 203, in dispatch
    return cls()._dispatch(*args, **kwargs)
  File "/home/kbalk/plus3it/moto/moto/core/responses.py", line 313, in _dispatch
    return self.call_action()
  File "/home/kbalk/plus3it/moto/moto/core/responses.py", line 403, in call_action
    response = method()
  File "/home/kbalk/plus3it/moto/moto/cloudformation/responses.py", line 264, in describe_stacks
    return template.render(stacks=stacks_resp, next_token=next_token)
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 1, in top-level template code
    
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/site-packages/jinja2/environment.py", line 471, in getattr
    return getattr(obj, attribute)
  File "/home/kbalk/plus3it/moto/moto/cloudformation/models.py", line 325, in stack_outputs
    return [v for v in self.output_map.values() if v]
  File "/home/kbalk/plus3it/moto/moto/cloudformation/models.py", line 325, in <listcomp>
    return [v for v in self.output_map.values() if v]
  File "/home/kbalk/.pyenv/versions/3.8.6/lib/python3.8/_collections_abc.py", line 762, in __iter__
    yield self._mapping[key]
  File "/home/kbalk/plus3it/moto/moto/cloudformation/parsing.py", line 810, in __getitem__
    new_output = parse_output(
  File "/home/kbalk/plus3it/moto/moto/cloudformation/parsing.py", line 409, in parse_output
    output_json = clean_json(output_json, resources_map)
  File "/home/kbalk/plus3it/moto/moto/cloudformation/parsing.py", line 205, in clean_json
    cleaned_val = clean_json(value, resources_map)
  File "/home/kbalk/plus3it/moto/moto/cloudformation/parsing.py", line 134, in clean_json
    return clean_json(true_value, resources_map)
  File "/home/kbalk/plus3it/moto/moto/cloudformation/parsing.py", line 175, in clean_json
    fn_sub_value = fn_sub_value.replace(sub, cleaned_ref)
TypeError: replace() argument 2 must be str, not None
```

**Comment by bblommers** [comment]
Re the external env variable, I've had similar setups in the past where the test framework would set the variable (I.e. in `tests/__init__.py`, or whatever makes sense for your setup).
That way, the Lambda can doesn't care about which mocking framework is used (Moto/Localstack/etc):
```
def handler:
    kwargs = dict()
    if "PROJECT_ENDPOINT_URL" in os.environ:
        kwargs["endpoint_url"] = os.environ["PROJECT_ENDPOINT_URL"]
    boto3.client("service", kwargs)
```

Not sure how where the error trace comes from though. The only external touchpoint in `settings.py` is a curl-command to http://docker, that should not be intercepted by Moto.

**Comment by kbalk** [comment]
The problem is that we're using Terraform scripts to test Terraform modules.  So there are two levels of Terraform variables.  We use `tftest` to kick off Terraform init/apply in either an actual or mock AWS environment,  so that's another place where I could set the environment variable, but I'd only know whether it's mock or not, and not whether it's mock docker or not.  There are just a bunch of layers where we'd have to pass that info around.  It would be nice to have that info in the lambda where it's actually used and needed.

**Comment by kbalk** [comment]
The stack trace is due to the fact that I forgot to add moto to the requirements file, so the lambda was probably crashing because it couldn't find the module to import.

**Comment by kbalk** [comment]
`moto_server_port()` returns a value of 5000, when the actual port we're using is 4615 (LocalStack starts moto_server from `initaws.d/start.sh`).
`moto_server_host()` returns "http://host.docker.internal:5000".

**Comment by kbalk** [comment]
Actually, the bigger problem is executing the lambda when LocalStack and moto are running in a docker container.  The message I see in the LogEvent using `localhost:4615/moto-api/#` is:

```
message: error running docker: 
409 Client Error for http+docker://localhost/v1.41/volumes/305eeed55e4383cb35437b2a6c91441fdb701a58ed35b6a73e72ad8ac0f640e3: 
Conflict ("remove 305eeed55e4383cb35437b2a6c91441fdb701a58ed35b6a73e72ad8ac0f640e3: 
volume is in use - [52bdb16fb33e9224edbc917a88b4e30d9dcf8a7cdbc3a6227d97e0e892bd3f64]")
```

**Comment by kbalk** [comment]
Some of these issues are particular to our test environment, but it would be nice to see the equivalent of LOCALSTACK_HOSTNAME.

**Comment by bblommers** [comment]
>  `moto_server_host()` returns "http://host.docker.internal:5000".

The fact that it returns `host.docker.internal` is slightly concerning - it should always return an IP-address inside Docker.

The fact that LS uses a different port means that we need to look at Docker's portbindings.

Could you try the following branch, and see what happens in for your setup?
https://github.com/bblommers/moto/tree/debug/moto-docker-ports

 - Exposes an environment variable called `MOTO_HOST` (with both host + port)
 - The external Docker port is used, e.g. 4615 (if available)
 - Adds some debug logging when trying to retrieve the ports

I'm especially interested in the debug logging - hopefully that gives some insight into why `host.docker.internal` is used

**Comment by kbalk** [comment]
For this to be tested, I'd need the `requirements.txt` file for the lambda to pull in your version of moto and for the docker container we use for running localstack and moto to pull in your version as well.  This may take me a little bit to figure out.

**Comment by kbalk** [comment]
I built a moto wheel and confirmed that it was being used when the lambda ran.  For an initial test, I started up LocalStack and moto from the command line, then ran the Terraform test.  `MOTO_PORT` is not set, `moto_server_port()` is still returning 5000, I see the message "Unable to get the host from the Docker API" and then I get a stack trace.

Here's the snippet I added to the lambda:

```
from moto.settings import moto_server_port, moto_server_host, is_docker, get_docker_host
if os.environ.get("MOCK_AWS"):   # This is a terraform variable I've set

    log.info("************************ klb stuff *******************************")
    moto_port = os.environ.get("MOTO_PORT", "")
    log.info("Value of moto_port: %s", moto_port)

    log.info("moto_server_port: %s", moto_server_port())
    log.info("moto_server_host: %s", moto_server_host())
    log.info("is_docker: %s", is_docker())
    log.info("get_docker_host: %s", get_docker_host())

    log.info("************************ end stuff *******************************")
```

Here's the log:  (Ignore the "BBB New code" messages ... I added that print statement in moto_server_port())

```
event_id: 0 	ingestion_time: 1638409283301 	message: ************************ klb stuff ******************************* 	timestamp: 1638409283301.6528
event_id: 1 	ingestion_time: 1638409283301 	message: Value of moto_port: 	timestamp: 1638409283301.6619
event_id: 2 	ingestion_time: 1638409283301 	message: ******** BBB New code ******* 	timestamp: 1638409283301.664
event_id: 3 	ingestion_time: 1638409283301 	message: ******** BBB New code ******* 	timestamp: 1638409283301.665
event_id: 4 	ingestion_time: 1638409283301 	message: 2021-12-02T01:41:22.581Z [lambda][INFO ]: ************************ klb stuff ******************************* 	timestamp: 1638409283301.667
event_id: 5 	ingestion_time: 1638409283301 	message: 2021-12-02T01:41:22.581Z [lambda][INFO ]: Value of moto_port: 	timestamp: 1638409283301.668
event_id: 6 	ingestion_time: 1638409283301 	message: 2021-12-02T01:41:22.581Z [lambda][INFO ]: moto_server_port: 5000 	timestamp: 1638409283301.669
event_id: 7 	ingestion_time: 1638409283301 	message: Unable to get host from Docker API: 	timestamp: 1638409283301.6711
event_id: 8 	ingestion_time: 1638409283301 	message: [32mSTART RequestId: b7f7d2a4-046d-1b62-c493-5f7613fe9c52 Version: $LATEST[0m 	timestamp: 1638409283301.6719
event_id: 9 	ingestion_time: 1638409283301 	message: [32mEND RequestId: b7f7d2a4-046d-1b62-c493-5f7613fe9c52[0m 	timestamp: 1638409283301.6729
event_id: 10 	ingestion_time: 1638409283301 	message: [32mREPORT RequestId: b7f7d2a4-046d-1b62-c493-5f7613fe9c52 Init Duration: 959.16 ms Duration: 3.32 ms Billed Duration: 4 ms Memory Size: 128 MB Max Memory Used: 29 MB [0m 	timestamp: 1638409283301.675
event_id: 11 	ingestion_time: 1638409283301 	message: [ERROR] AttributeError: 'JSONDecodeError' object has no attribute 'message' 	timestamp: 1638409283301.676
event_id: 12 	ingestion_time: 1638409283301 	message: Traceback (most recent call last): 	timestamp: 1638409283301.677
event_id: 13 	ingestion_time: 1638409283301 	message:   File "/var/lang/lib/python3.7/imp.py", line 234, in load_module 	timestamp: 1638409283301.678
event_id: 14 	ingestion_time: 1638409283301 	message:     return load_source(name, filename, file) 	timestamp: 1638409283301.6802
event_id: 15 	ingestion_time: 1638409283301 	message:   File "/var/lang/lib/python3.7/imp.py", line 171, in load_source 	timestamp: 1638409283301.6812
event_id: 16 	ingestion_time: 1638409283301 	message:     module = _load(spec) 	timestamp: 1638409283301.682
event_id: 17 	ingestion_time: 1638409283301 	message:   File "<frozen importlib._bootstrap>", line 696, in _load 	timestamp: 1638409283301.6838
event_id: 18 	ingestion_time: 1638409283301 	message:   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked 	timestamp: 1638409283301.685
event_id: 19 	ingestion_time: 1638409283301 	message:   File "<frozen importlib._bootstrap_external>", line 728, in exec_module 	timestamp: 1638409283301.686
event_id: 20 	ingestion_time: 1638409283301 	message:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed 	timestamp: 1638409283301.688
event_id: 21 	ingestion_time: 1638409283301 	message:   File "/var/task/lambda.py", line 54, in <module> 	timestamp: 1638409283301.689
event_id: 22 	ingestion_time: 1638409283301 	message:     log.info("moto_server_host: %s", moto_server_host()) 	timestamp: 1638409283301.6902
event_id: 23 	ingestion_time: 1638409283301 	message:   File "/var/task/moto/settings.py", line 56, in moto_server_host 	timestamp: 1638409283301.6912
event_id: 24 	ingestion_time: 1638409283301 	message:     return get_docker_host() 	timestamp: 1638409283301.692
event_id: 25 	ingestion_time: 1638409283301 	message:   File "/var/task/moto/settings.py", line 83, in get_docker_host 	timestamp: 1638409283301.6938
event_id: 26 	ingestion_time: 1638409283301 	message:     print(e.message) 	timestamp: 1638409283301.695
event_id: 27 	ingestion_time: 1638409283301 	message: 	timestamp: 1638409283301.696
event_id: 28 	ingestion_time: 1638409283301 	message: {"errorType":"AttributeError","errorMessage":"'JSONDecodeError' object has no attribute 'message'","stackTrace":[" File \"/var/lang/lib/python3.7/imp.py\", line 234, in load_module\n return load_source(name, filename, file)\n"," File \"/var/lang/lib/python3.7/imp.py\", line 171, in load_source\n module = _load(spec)\n"," File \"\u003cfrozen importlib._bootstrap\u003e\", line 696, in _load\n"," File \"\u003cfrozen importlib._bootstrap\u003e\", line 677, in _load_unlocked\n"," File \"\u003cfrozen importlib._bootstrap_external\u003e\", line 728, in exec_module\n"," File \"\u003cfrozen importlib._bootstrap\u003e\", line 219, in _call_with_frames_removed\n"," File \"/var/task/lambda.py\", line 54, in \u003cmodule\u003e\n log.info(\"moto_server_host: %s\", moto_server_host())\n"," File \"/var/task/moto/settings.py\", line 56, in moto_server_host\n return get_docker_host()\n"," File \"/var/task/moto/settings.py\", line 83, in get_docker_host\n print(e.message)\n"]} 	timestamp: 1638409283301.698
```

**Comment by bblommers** [comment]
OK, that gives a bit more context. I was confused how LocalStack and Moto were working together.
If Moto is started as a plain service, then the host should in fact be `host.docker.internal`.

The fact that there is a 'JsonDecodeError' makes it look like Moto thinks it is in fact inside a DockerContainer - but when trying to talk to the Docker API it fails (or at least, returns something that is not JSON).
That is a bit confusing.

Why the port is not picked up.. No idea. That bit of code should be the most straightforward.
I'll simplify the logic a bit more, and add some more debugging statements, at least before the weekend.

**Comment by bblommers** [comment]
> 
> This terraform test code is open source and I could tell you how I'm running it, but that doesn't seem fair to you. I've been transferred to another group so I'll probably have to wait until the weekend when I can spend more time digging into this. I think the issue is that the Lambda is running in its own silo, ignorant of the moto process that started it.

Well, I make no promises.. but if is easy to set-up, let me know, and I'll see if I have time.

**Comment by kbalk** [comment]
Note that the scenario where I run LocalStack and moto from the command line is the easier setup.  I thought I would try your changes in that environment as I can get the Lambda to run.

The setup where I can not get the Lambda to run at all is where LocalStack is run in a docker container.  It starts up moto using their `initaws.d` mechanism:

```
$ more initaws.d/start.sh 
#!/usr/bin/sh
nohup "$MAVEN_CONFIG"/.venv/bin/moto_server --host 0.0.0.0 --port 4615 &
echo "Pausing for moto to warm up ..." >&2
sleep 2
```

The test environment in that docker container includes the logic that creates the provider file.  The Lambda fails to start up and I get the volume busy message that I listed in a previous comment.

**Comment by kbalk** [comment]
I was looking at what LocalStack does and they have similar functions to determine if they're in a docker container or not:  https://github.com/localstack/localstack/blob/ce71b3f572b77ffe388d85034a853a1887d5645e/localstack/config.py#L559-L633

**Comment by bblommers** [comment]
> I'll simplify the logic a bit more, and add some more debugging statements, at least before the weekend.
>

Done.

> I was looking at what LocalStack does and they have similar functions to determine if they're in a docker container or not: 
>
Nice find - that is some very detailed logic.

**Comment by kbalk** [comment]
Got in a run this evening ....

```
$ moto_server --host 0.0.0.0 --port 4615
===========
is_docker?
DockerEnv exists: False
Path exists: True
============
4615
http://host.docker.internal:4615
 * Running on http://0.0.0.0:4615/ (Press CTRL+C to quit)
     ... <snip> ...
127.0.0.1 - - [02/Dec/2021 22:07:11] "POST / HTTP/1.1" 200 -
127.0.0.1 - - [02/Dec/2021 22:07:42] "POST /2015-03-31/functions HTTP/1.1" 201 -
127.0.0.1 - - [02/Dec/2021 22:07:47] "GET /2015-03-31/functions/test-cfn-lookup-ami-ids-54 HTTP/1.1" 200 -
127.0.0.1 - - [02/Dec/2021 22:07:47] "GET /2015-03-31/functions/test-cfn-lookup-ami-ids-54 HTTP/1.1" 200 -
127.0.0.1 - - [02/Dec/2021 22:07:47] "GET /2015-03-31/functions/test-cfn-lookup-ami-ids-54/versions?MaxItems=10000 HTTP/1.1" 200 -
is_docker?
DockerEnv exists: False
Path exists: True
   ... <snip> ...
```

The logs from the lambda itself show:

```
event_id: 13 	ingestion_time: 1638500877489 	message: 	timestamp: 1638500877489.654
event_id: 14 	ingestion_time: 1638500877489 	message: null 	timestamp: 1638500877489.655
event_id: 15 	ingestion_time: 1638501093841 	message: is_docker? 	timestamp: 1638501093841.383
event_id: 16 	ingestion_time: 1638501093841 	message: DockerEnv exists: True 	timestamp: 1638501093841.388
event_id: 17 	ingestion_time: 1638501093841 	message: Path exists: True 	timestamp: 1638501093841.389
event_id: 18 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.819Z [lambda][INFO ]: ************************ klb stuff ******************************* 	timestamp: 1638501093841.391
event_id: 19 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.819Z [lambda][INFO ]: Value of moto_port: 	timestamp: 1638501093841.392
event_id: 20 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.819Z [lambda][INFO ]: moto_server_port: 5000 	timestamp: 1638501093841.393
event_id: 21 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.902Z [lambda][INFO ]: moto_server_host: http://host.docker.internal:5000 	timestamp: 1638501093841.394
event_id: 22 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.902Z [lambda][INFO ]: is_docker: True 	timestamp: 1638501093841.395
event_id: 23 	ingestion_time: 1638501093841 	message: 	timestamp: 1638501093841.397
event_id: 24 	ingestion_time: 1638501093841 	message: Unable to get host from Docker API: 	timestamp: 1638501093841.398
event_id: 25 	ingestion_time: 1638501093841 	message: Expecting value: line 1 column 1 (char 0) 	timestamp: 1638501093841.399
event_id: 26 	ingestion_time: 1638501093841 	message: is_docker? 	timestamp: 1638501093841.4
event_id: 27 	ingestion_time: 1638501093841 	message: DockerEnv exists: True 	timestamp: 1638501093841.4011
event_id: 28 	ingestion_time: 1638501093841 	message: Path exists: True 	timestamp: 1638501093841.402
event_id: 29 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.921Z [lambda][INFO ]: get_docker_host: http://host.docker.internal:5000 	timestamp: 1638501093841.403
event_id: 30 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.921Z [lambda][INFO ]: ************************ end stuff ******************************* 	timestamp: 1638501093841.405
event_id: 31 	ingestion_time: 1638501093841 	message: 	timestamp: 1638501093841.406
event_id: 32 	ingestion_time: 1638501093841 	message: Unable to get host from Docker API: 	timestamp: 1638501093841.407
event_id: 33 	ingestion_time: 1638501093841 	message: Expecting value: line 1 column 1 (char 0) 	timestamp: 1638501093841.409
event_id: 34 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:32.947Z [botocore.credentials][INFO ]: Found credentials in environment variables. 	timestamp: 1638501093841.41
event_id: 35 	ingestion_time: 1638501093841 	message: [32mSTART RequestId: 10c3b51d-75b7-1e9a-a2a2-9ef8e7f2e846 Version: $LATEST[0m 	timestamp: 1638501093841.4111
event_id: 36 	ingestion_time: 1638501093841 	message: http://host.docker.internal:4615/cloudformation_us-east-1/cfnresponse?stack=arn:aws:cloudformation:us-east-1:123456789:stack/ami-lookup-test-complex-search/7f615cad-15e2-4068-9583-d976ca9756e8 	timestamp: 1638501093841.412
event_id: 37 	ingestion_time: 1638501093841 	message: Response body: 	timestamp: 1638501093841.413
event_id: 38 	ingestion_time: 1638501093841 	message: {"Status": "SUCCESS", "Reason": "See the details in CloudWatch Log Stream: 2021/12/03/[$LATEST]25031a0c8dd2839b0350377e41f3e472", "PhysicalResourceId": "2021/12/03/[$LATEST]25031a0c8dd2839b0350377e41f3e472", "StackId": "arn:aws:cloudformation:us-east-1:123456789:stack/ami-lookup-test-complex-search/7f615cad-15e2-4068-9583-d976ca9756e8", "RequestId": "3f7436fa-c117-403c-bff3-d8be1f97244e", "LogicalResourceId": "AmiIdLookup", "NoEcho": false, "Data": {"Data": {"Filters": "[{\"Name\":\"name\",\"Values\":[\"amzn-ami-hvm-*\"]},{\"Name\":\"description\",\"Values\":[\"Amazon Linux AMI*\"]},{\"Name\":\"architecture\",\"Values\":[\"x86_64\"]},{\"Name\":\"root-device-type\",\"Values\":[\"ebs\"]}]", "Owners": "[\"amazon\"]", "Region": "us-east-1", "ServiceToken": "arn:aws:lambda:us-east-1:123456789012:function:test-cfn-lookup-ami-ids-bd"}, "Id": "ami-760aaa0f"}} 	timestamp: 1638501093841.414
event_id: 39 	ingestion_time: 1638501093841 	message: 2021-12-03T03:11:33.062Z [lambda][INFO ]: found ami: ami-760aaa0f 	timestamp: 1638501093841.4158
event_id: 40 	ingestion_time: 1638501093841 	message: Status code: OK 	timestamp: 1638501093841.417
```

**Comment by kbalk** [comment]
The detailed logic in LocalStack hinges on LOCALSTACK_HOSTNAME.

**Comment by bblommers** [comment]
Well, that was less useful than I hoped for. So the Docker API call returns an empty string, giving us no information about hosts/ports whatsoever.

Are you on Windows, by any chance? Apparently the Docker API is not available in Windows by default.

**Comment by kbalk** [comment]
I'm running on Linux, a Centos 7 system.

**Comment by bblommers** [comment]
Ah, that actually has the same issue - the Docker API isn't available on CentOS by default, only after some config changes.

I can't really think of a feasible way for Moto to determine the port, if the Docker API is not available. (Other than using the env variables...)

**Comment by kbalk** [comment]
Are you completely against passing something like the hostname and port into the lambda's environment?

I was planning on perusing LocalStack's code to see how they were handling LOCALSTACK_HOSTNAME and how they were handling the docker container.  I also need to look at moto's code for handling the docker container.  I figured once I did that, I would be in a better position to see what's possible.

**Comment by bblommers** [comment]
> Are you completely against passing something like the hostname and port into the lambda's environment?
> 

No, not at all - just pointing out that I can't see how Moto would determine the correct values. So the user would always have to set the env vars themselves when starting Moto, and Moto then propagates them onto the lambda container.

> I was planning on perusing LocalStack's code to see how they were handling LOCALSTACK_HOSTNAME and how they were handling the docker container. I also need to look at moto's code for handling the docker container. I figured once I did that, I would be in a better position to see what's possible.
>

https://github.com/spulec/moto/blob/master/moto/awslambda/models.py#L588:L598

One of the additional options we could set is `network_mode="host"`. That would allow us to always use `localhost` to refer to the host system. It would be a breaking change though, so it would have to be part of a major release.
(Note that it still wouldn't tell us the port, so that would always have to be set as an env var.)

**Comment by kbalk** [comment]
Would you consider something like this:

```
diff --git a/moto/awslambda/models.py b/moto/awslambda/models.py
index 6dc49323..a91ed850 100644
--- a/moto/awslambda/models.py
+++ b/moto/awslambda/models.py
@@ -568,6 +568,8 @@ class LambdaFunction(CloudFormationModel, DockerModel):
             }
 
             env_vars.update(self.environment_vars)
+            env_vars["MOTO_HOST"] = settings.moto_server_host()
+            env_vars["MOTO_PORT"] = settings.moto_server_port()
```

I could then retrieve those two environment variables from the lambda.  Those environment variables would need to be documented so people could take advantage of them.  It's not much different than what LocalStack is doing with LOCALSTACK_HOSTNAME.

I think there's a different problem (outside of determining the endpoint) when LocalStack and Moto are running in a docker container.  I'm having trouble getting that environment to run at the moment because of the wheel I created.  I'll work on that separately.  I wasted so much time earlier in the day due to a very stupid typo.

**Comment by lorengordon** [comment]
>One of the additional options we could set is `network_mode="host"`. That would allow us to always use `localhost` to refer to the host system. It would be a breaking change though, so it would have to be part of a major release.

How about allowing the user to specify a value for the `network` option, somehow, instead? Maybe a cli option in server mode, or an env? I work with Karri, and I suspect part of the difficulty we've been having is that we specify a user-defined bridge when starting the container that is running both localstack and moto server... The moto code currently does not set the network when invoking the lambda, which I believe means that container is using only the default bridge.

https://github.com/spulec/moto/blob/2fde6be849923b89a2e76de6c3a8221ed8cd9cf3/moto/awslambda/models.py#L591-L601

**Comment by bblommers** [comment]
That's an option as well, @lorengordon. Do you think you would need a different network-mode than `host`, or are you suggesting an env as a way to make this a non-breaking change?

**Comment by lorengordon** [comment]
I'm suggesting it as a non-breaking change. Do no specify `network_mode` so that it continues to use bridge mode, but allow the user to optionally set `network=<value>` to connect to a user-defined bridge.

Something like:

```
 container = self.docker_client.containers.run( 
     image_ref, 
     [self.handler, json.dumps(event)], 
     remove=False, 
     mem_limit="{}m".format(self.memory_size), 
     volumes=["{}:/var/task".format(data_vol.name)], 
     environment=env_vars, 
     network=settings.moto_lambda_network(), # where this is `"bridge"` by default, or the user-specified value
     detach=True, 
     log_config=log_config, 
     **run_kwargs 
 ) 
```

Or maybe update `run_kwargs` with the `network` key if the user provided a value.

**Comment by bblommers** [comment]
This is now all available as of moto >= 3.0.2.dev7. 
The documentation on all of the configuration options can be found here: http://docs.getmoto.org/en/latest/docs/services/lambda.html

**Comment by lorengordon** [comment]
Amazing, this is an awesome feature! Thanks @bblommers!
